File: python3/Dynamic Programming/maximum_subarray_sum.py
from typing import List


def maximum_subarray_sum(nums: List[int]) -> int:
    if not nums:
        return 0
    # Set the sum variables to negative infinity to ensure negative 
    # sums can be considered.
    max_sum = current_sum = float('-inf')
    # Iterate through the array to find the maximum subarray sum.
    for num in nums:
        # Either add the current number to the existing running sum, or 
        # start a new subarray at the current number.
        current_sum = max(current_sum + num, num)
        max_sum = max(max_sum, current_sum)
    return max_sum

File: cpp/Dynamic Programming/maximum_subarray_sum.cpp
#include <vector>
#include <algorithm>
#include <limits>

int maximumSubarraySum(std::vector<int>& nums) {
    if (nums.empty()) {
        return 0;
    }
    // Set `maxSum` to negative infinity to ensure negative
    // sums can be considered.
    int maxSum = std::numeric_limits<int>::min();
    int currentSum = 0;
    // Iterate through the array to find the maximum subarray sum.
    for (int num : nums) {
        // Either add the current number to the existing running sum, or 
        // start a new subarray at the current number.
        currentSum = std::max(currentSum + num, num);
        maxSum = std::max(maxSum, currentSum);
    }
    return maxSum;
}